flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java

```diff
@@ -94,26 +130,218 @@ public RegistrationResponse apply(final JobMasterGateway jobMasterGateway) {
+					// save the register taskExecutor gateway
+					taskExecutorGateways.put(resourceID, taskExecutorGateway);
+					// schedule the heartbeat with the registered taskExecutor
+					ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatScheduler = new ResourceManagerToTaskExecutorHeartbeatScheduler(
+						ResourceManager.this, leaderSessionID, taskExecutorGateway, taskExecutorAddress, resourceID, log);
+					heartbeatScheduler.start();
+					heartbeatSchedulers.put(resourceID, heartbeatScheduler);
+					return new TaskExecutorRegistrationSuccess(new InstanceID(), heartbeatScheduler.getHeartbeatInterval());
+				}
+			}
+		}, getMainThreadExecutionContext());
+
+	}
+
+	/**
+	 * notify resource failure to resourceManager, because of two reasons:
+	 * 1. cannot keep heartbeat with taskManager for several times, mark the resource as failed
+	 * 2. in some corner cases, TM will be marked as invalid by cluster manager master(e.g. yarn master), but TM itself does not realize.
+	 *
+	 * @param resourceID identify the taskManager which to stop
+	 */
+	@RpcMethod
+	public void notifyResourceFailure(ResourceID resourceID) {
+		log.warn("receive failure notification of resource {}", resourceID);
+		TaskExecutorGateway taskManager = taskExecutorGateways.get(resourceID);
+		if(taskManager == null) {
+			// ignore command to stop an unregister taskManager
+			log.warn("ignore stop taskManager command because {} is unregistered", resourceID);
+		} else {
+			taskExecutorGateways.remove(resourceID);
+			closeHeartbeatToResourceIfExist(resourceID);
+			// TODO notify slotManager and notify jobMaster,
+			// TODO slotManager.notifyTaskManagerFailure(resourceID);
+			taskManager.shutDown(leaderSessionID);
+		}
+	}
+
+	/**
+	 * close heartbeat triggers to resource if exist
+	 * @param resourceID which resource need to stop keep heartbeat with
+	 */
+	private void closeHeartbeatToResourceIfExist(ResourceID resourceID) {
+		if(heartbeatSchedulers.containsKey(resourceID)) {
+			ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatManager = heartbeatSchedulers.get(resourceID);
+			heartbeatManager.close();
+			heartbeatSchedulers.remove(resourceID);
+		}
+	}
+
+	/**
+	 * send slotRequest to the taskManager which the given slot is on
+	 *
+	 * @param slotRequest slot request information
+	 * @param slotID      which slot is choosen
+	 */
+	void requestSlotToTaskManager(final SlotRequest slotRequest, final SlotID slotID) {
+		ResourceID resourceID = slotID.getResourceID();
+		TaskExecutorGateway taskManager = taskExecutorGateways.get(resourceID);
+		if (taskManager == null) {
+			// the given slot is on an unregister taskManager
+			log.warn("ignore slot {} because it is on an unregister taskManager", slotID);
+			// TODO slotManager.handleSlotRequestFailedAtTaskManager(slotRequest, slotID);
+		} else {
+			Future<SlotAllocationResponse> response = taskManager.requestSlotForJob(
+				slotRequest.getAllocationID(), slotRequest.getJobID(), slotID, leaderSessionID);
```

Comment:
Why do you unwrap the `SlotRequest` here? We could also send the `SlotRequest` to the `TaskExecutor`.